### PR TITLE
Handle functools.partial/partialmethod in validators

### LINKED ISSUE #5352: Remove requirement that callables have `__qualname__` attribute

### PR DESCRIPTION
**Pull request by adriangb**
Fixes #5352

Selected Reviewer: @samuelcolvin

### CODE REVIEW

**Comment by adriangb** [comment]
Missing serializers (WIP) but please review

File: pydantic/_internal/_decorators.py [hunk 1]
```diff
@@ -341,66 +349,93 @@ def gather_decorator_functions(cls: type[Any]) -> DecoratorInfos:
...
+def prepare_serializer_decorator(function: AnyDecoratorCallable, allow_reuse: bool) -> Any:
     """
     Warn about validators/serializers with duplicated names since without this, they can be overwritten silently
     which generally isn't the intended behaviour, don't run in ipython (see #312) or if `allow_reuse` is True.
     """
-    if isinstance(function, staticmethod):
-        function = function.__func__
     if not allow_reuse and not in_ipython():
-        ref = f'{function.__module__}::{function.__qualname__}'
+        ref = get_function_ref(function)
```
Comment (dmontagu) [inline pydantic/_internal/_decorators.py:398]: In `check_for_duplicate_decorator_function ` below you do:
```python
        function = unwrap_wrapped_function(function)
        ref = get_function_ref(function)
```
Whereas here, you don't do the unwrapping. Is this intentional? Without digging into the full logic path it seems like it should be the same in both places but I could imagine it being possible you already know the function has been unwrapped here, or similar.

Comment (adriangb) [reply pydantic/_internal/_decorators.py]: I just deleted this function _and_ deleted the `function = unwrap_wrapped_function(function)` in `check_for_duplicate_decorator_function` because `get_function_ref` already does that.

**Comment by adriangb** [comment]
please review

File: pydantic/_internal/_decorators.py [hunk 2]
```diff
@@ -341,68 +349,83 @@ def gather_decorator_functions(cls: type[Any]) -> DecoratorInfos:
...
-    """
-    if isinstance(function, staticmethod):
-        function = function.__func__
-    if not allow_reuse and not in_ipython():
-        ref = f'{function.__module__}::{function.__qualname__}'
-        if ref in _FUNCS:
-            warnings.warn(f'duplicate validator function "{ref}"; if this is intended, set `allow_reuse=True`')
-        _FUNCS.add(ref)
-    return function
+AnyDecoratorCallable: TypeAlias = 'Union[ classmethod[Any], staticmethod[Any], partialmethod[Any], Callable[..., Any]]'
```
Comment (samuelcolvin) [inline pydantic/_internal/_decorators.py:352]:
```suggestion
AnyDecoratorCallable: TypeAlias = 'Union[classmethod[Any], staticmethod[Any], partialmethod[Any], Callable[..., Any]]'
```

File: pydantic/_internal/_decorators.py [hunk 3]
```diff
@@ -341,68 +349,83 @@ def gather_decorator_functions(cls: type[Any]) -> DecoratorInfos:
...
-    Unwrap unbound classmethods and staticmethods
-    """
-    if isinstance(function, (classmethod, staticmethod)):
-        return function.__func__
-    return function
+def unwrap_wrapped_function(
+    func: Any,
+    *,
+    unwrap_class_static_method: bool = True,
+) -> Any:
```
Comment (samuelcolvin) [inline pydantic/_internal/_decorators.py:359]: could we add a docstring explaining what this is doing.

File: pydantic/_internal/_decorators.py [hunk 4]
```diff
@@ -341,68 +349,83 @@ def gather_decorator_functions(cls: type[Any]) -> DecoratorInfos:
...
+            staticmethod,
+            classmethod,
+            partial,
+            partialmethod,
+        )
+    else:
+        all = (
+            partial,
+            partialmethod,
+        )
```
Comment (samuelcolvin) [inline pydantic/_internal/_decorators.py:372]:
```suggestion
        all = partial, partialmethod
```

File: pydantic/_internal/_decorators.py [hunk 5]
```diff
@@ -341,68 +349,83 @@ def gather_decorator_functions(cls: type[Any]) -> DecoratorInfos:
...
     Warn about validators with duplicated names since without this, they can be overwritten silently
     which generally isn't the intended behaviour, don't run in ipython (see #312) or if `allow_reuse` is True.
     """
     if not allow_reuse and not in_ipython():
-        function = unwrap_unbound_methods(function)
-        ref = f'{function.__module__}::{function.__qualname__}'
+        ref = get_function_ref(function)
         if ref in _FUNCS:
-            warnings.warn(f'duplicate validator function "{ref}"; if this is intended, set `allow_reuse=True`')
+            warnings.warn(f'duplicate {type} function "{ref}"; if this is intended, set `allow_reuse=True`')
```
Comment (samuelcolvin) [inline pydantic/_internal/_decorators.py:428]: Remember we want to replace this with #4436.

But happy to defer that to a new PR. @hramezani, probably don't start work on #4436 until this is merged.

**Review by samuelcolvin — COMMENTED** [review]
otherwise I think looks good.

**Comment by samuelcolvin** [comment]
also conflicts, please update.

File: pydantic/_internal/_decorators.py (hunk 3 above)
Comment (adriangb) [reply pydantic/_internal/_decorators.py:376]: done

**Comment by adriangb** [comment]
please review

**Comment by hramezani** [comment]
Maybe not related to this PR, but I can't see any test for duplicate serializers function like [the test_duplicate for validator](https://github.com/pydantic/pydantic/blob/0cd57b3e0c75d9622012de542e5ddc110fe55743/tests/test_validators.py#L370). I think by the change of this PR, both serializers and validators are going to use `check_for_duplicate_decorator_function` but I think it worth to add some test for duplicate serializers. Also, the above-mentioned test is for `field_validator`, what about adding tests for other validators?

**Comment by samuelcolvin** [comment]
Good point, but let's leave that until we replace #4436.

**Comment by samuelcolvin** [comment]
great, thanks so much.
